[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123067
* __->__ #121625

**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented:
I've been a bit busy today, will try to have a look at this one tomorrow!

lezcano commented on torch/_inductor/scheduler.py:
```diff
@@ -998,6 +999,51 @@ def debug_str(self) -> str:
         return "\n".join(lines).rstrip()
 
 
+# TODO<Leslie>: Find a better name
+class OuterFusedSchedulerNode(FusedSchedulerNode):
```
This is a C++ construct, it should go in cpp.py. The other two uses of OuterFusedSchedulerNode in this file can be removed given that OuterFusedSchedulerNode inherits from FusedSchedulerNode.

leslie-fang-intel replied:
Thanks for the suggestion and changed.

lezcano reviewed (approved):
In a perfect world, we would be able to perform the pass where we figure out the tiling of each node before fusion. This would help up considering whether we want to fuse nodes that would disallow vectorisation / tiling. In that perfect world, we could move this loop_fusion_level to the fusion step of the codegen, but well. At the moment we are quite far from there, so I guess the current implementation is as good as it gets for now.

Thank you for the cool feature!

leslie-fang-intel commented on torch/_inductor/scheduler.py:
```diff
@@ -1675,22 +1721,47 @@ def fuse_nodes(self):
+            # Perform outer loop fusion for the left nodes with lower priority when backend support is available.
+            ctx_manager = (
+                self.get_backend(torch.device("cpu")).enable_outer_loop_fusion()
+                if (all_node_in_cpu and _check_outer_loop_fusion)
```
Thanks for the suggestion and changed.

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/metrics.py:
```diff
@@ -42,6 +42,10 @@
 # counters for tracking cpp_wrapper disabled
 disable_cpp_wrapper = 0
 
+# The length counts the number of loops fused at the outer loop.
+# Each element counts the number of inner loops for each fused outer loop.
+cpp_fused_outer_loop_nest: List[int] = []
```
```suggestion
# The length counts the number of outer loop fusions.
# Each element counts the number of inner kernels in each outer loop fusion.
cpp_outer_loop_fused_inner_counts: List[int] = []
```

leslie-fang-intel replied:
Thanks and changed.

leslie-fang-intel commented on torch/_inductor/scheduler.py:
```diff
@@ -2412,7 +2438,7 @@ def codegen_template(
         raise NotImplementedError()
 
-    def codegen_nodes(self, nodes: List[SchedulerNode]):
+    def codegen_nodes(self, node: Union[FusedSchedulerNode, SchedulerNode]):
```
Thanks for the comment and changed.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3608,19 +3756,154 @@ def can_fuse_horizontal(self, node1, node2):
+        if (
+            reduce1 != ()
+            and reduce2 == ()
+            and vars1 + reduce1 == vars2  # Reduction + ElementWise
```
Thanks and changed.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3608,19 +3756,154 @@ def can_fuse_horizontal(self, node1, node2):
-        cpp_kernel_proxy = CppKernelProxy(kernel_group)
-        cpp_kernel_proxy.codegen_nodes(nodes)
+        if isinstance(node, OuterFusedSchedulerNode):
+            lazy_cpp_kernel_proxy_list = []
```
Thanks and remove the `lazy` prefix.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -360,6 +362,51 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
     return stride_at(index_vec_simplified, var)
 
 
+# TODO<Leslie>: Find a better name
+class OuterFusedSchedulerNode(FusedSchedulerNode):
```
Thanks for the suggestion and changed.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3608,19 +3756,154 @@ def can_fuse_horizontal(self, node1, node2):
+            self._get_outer_loop_fusion_status()
+            and type(node1)
+            in (OuterFusedSchedulerNode, FusedSchedulerNode, SchedulerNode)
+            and type(node2) in (FusedSchedulerNode, SchedulerNode)
```
I think we can support the case when `type(node2)` is `OuterFusedSchedulerNode`. Changed as suggested.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -360,6 +362,51 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+            OuterFusedSchedulerNode,
+            SchedulerNode,
+            FusedSchedulerNode,
+        ) and type(node2) in (SchedulerNode, FusedSchedulerNode)
```
I think we can support the case when `type(node2)` is `OuterFusedSchedulerNode`. Changed as suggested.

leslie-fang-intel commented on torch/_inductor/scheduler.py:
```diff
@@ -1678,22 +1679,47 @@ def fuse_nodes(self):
+        )
+
+        for _check_outer_loop_fusion in check_outer_loop_fusion_list:
+            # Full loop fusion should be prioritized in the fusion process, as it typically leads to better performance.
+            # Perform outer loop fusion for the left nodes with lower priority when backend support is available.
```
The current [implementation of `fuse_nodes`](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/scheduler.py#L1681-L1696) conducts a for-loop 10 times to ensure that all `full fusions` are completed before moving on to the `outer fusion` passes. If we modify the implementation to extend the existing fusion scoring mechanism (`self.score_fusion`), where each for-loop encompasses both `full fusions` and `outer fusion`, there's a concern that we may miss opportunities for `full fusions` as nodes may be chosen by the `outer fusion` process.

lezcano replied:
In that case, you should just score these fusions quite low so that they never take preference over a regular fusion, right?

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3608,19 +3756,154 @@ def can_fuse_horizontal(self, node1, node2):
+        ) or (
+            reduce1 != ()
+            and reduce2 != ()
+            and vars1 + reduce1 == vars2 + reduce2  # Reduction + Reduction
```
Thanks for the suggestion. Changed it as synced offline.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3608,19 +3756,154 @@ def can_fuse_horizontal(self, node1, node2):
+                        "size",
+                        "offset",
+                        "steps",
+                        "kernel",
```
Remove this check.

leslie-fang-intel replied:
The check of `kernel` has been removed.

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3608,19 +3756,154 @@ def can_fuse_horizontal(self, node1, node2):
+                        "offset",
+                        "steps",
+                        "kernel",
+                        "reduction_var_map",
```
Thanks and removed.

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1828,7 +1891,61 @@ def gen_loop(loop: LoopLevel, in_reduction=False):
 
             stack.enter_context(code.indent())
-            if loop_nest.root:
+            if outer_loop_fusion:
```
I think this is related to https://github.com/pytorch/pytorch/pull/121064#issuecomment-1985489574, and https://github.com/pytorch/pytorch/pull/121625#pullrequestreview-1927398754, so I'm not sure whether it can be fixed without a large refactor.

jgong5 replied:
> I think this is related to [#121064 (comment)](https://github.com/pytorch/pytorch/pull/121064#issuecomment-1985489574), and [#121625 (review)](https://github.com/pytorch/pytorch/pull/121625#pullrequestreview-1927398754), so I'm not sure whether it can be fixed without a large refactor.

@leslie-fang-intel and I talked offline and think that it should be doable without a lot of refactoring, by reusing the existing `gen_loops` and `gen_kernel` based on the new `OuterLoopFusedKernel` abstraction. Let's see how it turns out.

leslie-fang-intel replied:
`OuterLoopFusedKernel` abstraction has been added. I will do some further code clean up.

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @lezcano @jgong5, thanks for your kind suggestions. I have made some refactor based on that.

- Add the `OuterLoopFusedKernel` abstract, so we can reuse most of the existing `gen_loops` and `gen_kernel` function.
- Most the util function like: `_check_outer_fusion_loop_level_attr` and `_merge_outer_fusion_kernels` into `OuterLoopFusedSchedulerNode`. It makes the flow in `codegen_node` clearer.

Please kindly help to take a look again.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+                _loop_level_split_num = len(_loop_level_nested_list[0])
+                for _loop_level_split_idx in range(_loop_level_split_num):
+                    # If at any outer loop fusion depth, we have split loop into main loop and tail loop,
+                    # The main loop and tail loop should have different OuterLoopFusedKernel.
+                    outer_loop_fused_kernel = OuterLoopFusedKernel()
```
Do I understand correctly that you are constructing the kernel from:
```
main_loop1:
    main_kernel1
tail_loop1:
    tail_kernel1
main_loop2:
    main_kernel2
tail_loop2:
    tail_kernel2
```
to
```
outer_fused_kernel1:
    main_loop1:
        main_kernel1
    main_loop2:
        main_kernel2
outer_fused_kernel2:
    tail_loop1:
        tail_kernel1
    tail_loop2:
        tail_kernel2
```
But shouldn't it be a single outer fused kernel:
```
outer_fused_kernel:
    main_loop1:
        main_kernel1
    tail_loop1:
        tail_kernel1
    main_loop2:
        main_kernel2
    tail_loop2:
        tail_kernel2
```

leslie-fang-intel replied:
As we discussed offline, here is used to support the case when loop split happens at any outer loop fusion depth. We have simplified this code by blocking the loop split at outer loop fusion depth in `check_outer_fusion_loop_level_attr`.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1828,18 +2027,43 @@ def codegen_loops_impl(self, loop_nest, code, worksharing):
             def gen_kernel(kernel):
-                with contextlib.ExitStack() as stack:
-                    assert kernel
+                if type(kernel) is OuterLoopFusedKernel:
```
Thanks and changed.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1828,18 +2027,43 @@ def codegen_loops_impl(self, loop_nest, code, worksharing):
+
+                kernels = loop.get_kernels()
+                assert len(kernels) == 1
+                if type(
+                    kernels[0]
+                ) is not OuterLoopFusedKernel and is_parallel_reduction(loop):
```
`gen_loop_kernel` will be used by both norm code gen and outer loop fusion code gen. I think we shouldn't assert not `OuterLoopFusedKernel` here.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+
+        return True
+
+    def _merge_outer_fusion_kernels(
```
Thanks and changed.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+            _outer_loop_fusion_depth,
+        ):
+            assert _outer_loop_fusion_depth >= 1
+            if (_outer_loop_fusion_depth := _outer_loop_fusion_depth - 1) >= 1:
```
Sure, will remove the leading underscores.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+    def get_outer_nodes(self):
+        return self.outer_fused_nodes
+
+    def _check_outer_fusion_loop_level_attr(
```
Thanks and changed.

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+                            ),
+                        )
+                    _loop_level_of_first_kernel.inner = []
+                    _loop_level_of_first_kernel.kernel = outer_loop_fused_kernel  # type: ignore[assignment]
```
Sure, make `OuterLoopFusedKernel` a subclass of `CppKernel.`

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+            ):
+                # Ensure same parallel depth.
+                # Refer to poolformer_m36 in timm_models:
+                # Node1 has loop collapsed(2), Node2 can't be loop collapsed due to it's reduction.
```
Thanks for the comment. This check has been removed. After we refactor to use the `OuterLoopFusedKernel`, this issue has gone. I think that's because [`max_parallel_depth`](https://github.com/pytorch/pytorch/blob/2b4173e0def678b04efea46a85909bb86f7c3fd6/torch/_inductor/codegen/cpp.py#L4087-L4090) only counts the loop level which has inners. After the refactor, it's same as the outer loop fusion depth.

leslie-fang-intel commented:
Hi @lezcano, since I have done some code refactor, could you please kindly help to take a look of this PR again?

lezcano commented:
Hey, sorry, I've been out the last 3 days, and I've been playing catch-up on reviews the whole day of today. I expect to get to this one tomorrow!

lezcano reviewed (commented):
Just one point, otherwise LGTM.

Note that, in the future, when every kernel can be vectorized, the function `select_tiling` will be much easier, and we'll be able to lift `check_outer_fusion_loop_level_attr` to the fusion phase, avoiding grouping kernels at fusion time that "look like we'll be able to fuse them" but then we actually won't.

leslie-fang-intel commented:
> In that case, you should just score these fusions quite low so that they never take preference over a regular fusion, right?

Thanks for the comment. If my understanding is right, `fuse_nodes_once` will fuse every pair of possible fusion nodes according to their scores, encompassing both regular fusions and outer loop fusions, within a single run of the function.

However, given that [`fuse_nodes`](https://github.com/pytorch/pytorch/blob/f4ff063c333f286d4384523bac67c047aca4d7b9/torch/_inductor/scheduler.py#L1679-L1698) will invoke `fuse_nodes_once` 10 times. I'm concerned that the `score approach` may potentially disrupt pairs of `regular fusion nodes` in the 2rd invocation of `fuse_nodes_once` if they have been pick up by the `outer loop fusion` in the 1st invocation of `fuse_nodes_once`.

leslie-fang-intel commented:
> Note that, in the future, when every kernel can be vectorized, the function select_tiling will be much easier, and we'll be able to lift check_outer_fusion_loop_level_attr to the fusion phase, avoiding grouping kernels at fusion time that "look like we'll be able to fuse them" but then we actually won't.

Add this into the comment of `check_outer_fusion_loop_level_attr`.

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [ReImplement] Outer Loop Fusion for CPP Backend"


**Summary**
Re-implement of https://github.com/pytorch/pytorch/pull/121064

**Test Plan**
```
python -u -m pytest -s -v test_cpu_repro.py -k test_outer_loop_fusion
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
